internal/mode/static/telemetry/collector.go

```diff
@@ -0,0 +1,156 @@
+	GetLatestConfiguration() *dataplane.Configuration
+}
+
+// NGFResourceCounts stores the counts of all relevant Graph resources.
+type NGFResourceCounts struct {
+	Gateways       int
+	GatewayClasses int
+	HTTPRoutes     int
+	Secrets        int
+	Services       int
+	Endpoints      int
+}
+
+// ProjectMetadata stores the name of the project and the current version.
+type ProjectMetadata struct {
+	Name    string
+	Version string
+}
+
+// Data is telemetry data.
+// Note: this type might change once https://github.com/nginxinc/nginx-gateway-fabric/issues/1318 is implemented.
+type Data struct {
+	ProjectMetadata   ProjectMetadata
+	NodeCount         int
+	NGFResourceCounts NGFResourceCounts
+}
+
+// DataCollectorConfig holds configuration parameters for DataCollectorImpl.
+type DataCollectorConfig struct {
+	// K8sClientReader is a Kubernetes API client Reader.
+	K8sClientReader client.Reader
+	// GraphGetter allows us to get the Graph.
+	GraphGetter GraphGetter
+	// ConfigurationGetter allows us to get the Configuration.
+	ConfigurationGetter ConfigurationGetter
+	// Version is the NGF version.
+	Version string
+}
+
+type DataCollectorImpl struct {
+	cfg DataCollectorConfig
+}
+
+func NewDataCollector(
+	cfg DataCollectorConfig,
+) *DataCollectorImpl {
+	return &DataCollectorImpl{
+		cfg: cfg,
+	}
+}
+
+// Collect collects and returns telemetry Data.
+func (c DataCollectorImpl) Collect(ctx context.Context) (Data, error) {
+	nodeCount, err := collectNodeCount(ctx, c.cfg.K8sClientReader)
+	if err != nil {
+		return Data{}, err
+	}
+
+	graphResourceCount, err := collectGraphResourceCount(c.cfg.GraphGetter, c.cfg.ConfigurationGetter)
+	if err != nil {
+		return Data{}, err
+	}
+
+	data := Data{
+		NodeCount:         nodeCount,
+		NGFResourceCounts: graphResourceCount,
+		ProjectMetadata: ProjectMetadata{
+			Name:    "NGF",
+			Version: c.cfg.Version,
+		},
+	}
+
+	return data, nil
+}
+
+func collectNodeCount(ctx context.Context, k8sClient client.Reader) (int, error) {
+	nodes := v1.NodeList{}
+	if err := k8sClient.List(ctx, &nodes); err != nil {
+		return 0, err
+	}
+
+	return len(nodes.Items), nil
+}
+
+func collectGraphResourceCount(
+	graphGetter GraphGetter,
+	configurationGetter ConfigurationGetter,
+) (NGFResourceCounts, error) {
+	ngfResourceCounts := NGFResourceCounts{}
+	g := graphGetter.GetLatestGraph()
+	cfg := configurationGetter.GetLatestConfiguration()
+
+	if g == nil {
+		return NGFResourceCounts{}, errors.New("latest graph cannot be nil")
+	}
+	if cfg == nil {
+		return NGFResourceCounts{}, errors.New("latest configuration cannot be nil")
+	}
+
+	if g.GatewayClass != nil {
+		ngfResourceCounts.GatewayClasses = 1
+	}
+	if g.Gateway != nil {
+		ngfResourceCounts.Gateways = 1
+	}
+	ngfResourceCounts.HTTPRoutes = len(g.Routes)
+	ngfResourceCounts.Secrets = countReferencedResources(g.ReferencedSecrets)
+	ngfResourceCounts.Services = countReferencedResources(g.ReferencedServices)
+
+	for _, upstream := range cfg.Upstreams {
```

Comment:
How about we simply report.  And calculate on the other side.
Do we really lose anything?  Does the precision matter?
Asking since calculating is overhead.

If there was something tied to licensing, I would say it totally matters.
But I have also seen 1 to 1200 reported by customers of NIC over time.

Reply:
We would be losing the data that would indicate the variation in endpoints between the 24hr send times. In my opinion, I don't think the variation within 24 hours is interesting enough to warrant this. We might find something surprising here, like the variation is sometimes really high within a day, but I don't think we'd use that information to prioritize any features.

I'd vote to keep it simple and report the "EndpointsPerUpstream" at the time we report, thus 'sampling" that data point once every 24 hours. With that, we can derive calculations on the other side that can answer more interesting questions, such as how fast are upstreams added when NGF installed? What are the percentiles, such as 95%, of total upstreams so we make sure we account for the upper bounds? etc

Reply:
Decided to keep total Endpoints but address this in issue #1528.